psi4/src/psi4/libdiis/diismanager.cc

```diff
@@ -213,53 +213,57 @@ bool DIISManager::add_entry(int numQuantities, ...) {
                 for (int h = 0; h < buf4->params->nirreps; ++h) {
                     global_dpd_->buf4_mat_irrep_init(buf4, h);
                     global_dpd_->buf4_mat_irrep_rd(buf4, h);
-                    for (int row = 0; row < buf4->params->rowtot[h]; ++row) {
-                        for (int col = 0; col < buf4->params->coltot[h]; ++col) {
-                            *arrayPtr++ = buf4->matrix[h][row][col];
-                        }
-                    }
+                    auto size = buf4->params->rowtot[h] * buf4->params->coltot[h ^ buf4->file.my_irrep];
+                    std::copy(buf4->matrix[h][0], buf4->matrix[h][0] + size, arrayPtr);
+                    arrayPtr += size;
                     global_dpd_->buf4_mat_irrep_close(buf4, h);
                 }
                 break;
+            }
             case DIISEntry::InputType::DPDFile2:
+            {
                 file2 = va_arg(args, dpdfile2 *);
                 global_dpd_->file2_mat_init(file2);
                 global_dpd_->file2_mat_rd(file2);
                 for (int h = 0; h < file2->params->nirreps; ++h) {
-                    for (int row = 0; row < file2->params->rowtot[h]; ++row) {
-                        for (int col = 0; col < file2->params->coltot[h]; ++col) {
-                            *arrayPtr++ = file2->matrix[h][row][col];
-                        }
-                    }
+                    auto size = file2->params->rowtot[h] * file2->params->coltot[file2->my_irrep ^ h];
+                    std::copy(file2->matrix[h][0], file2->matrix[h][0] + size, arrayPtr);
+                    arrayPtr += size;
                 }
                 global_dpd_->file2_mat_close(file2);
                 break;
+            }
             case DIISEntry::InputType::Matrix:
+            {
                 matrix = va_arg(args, Matrix *);
                 for (int h = 0; h < matrix->nirrep(); ++h) {
-                    for (int row = 0; row < matrix->rowspi()[h]; ++row) {
-                        for (int col = 0; col < matrix->colspi()[h]; ++col) {
-                            *arrayPtr++ = matrix->get(h, row, col);
-                        }
-                    }
+                    auto size = matrix->rowspi()[h] * matrix->colspi()[h ^ matrix->symmetry()];
+                    std::copy(matrix->pointer(h)[0], matrix->pointer(h)[0] + size, arrayPtr);
+                    arrayPtr += size;
                 }
                 break;
+            }
             case DIISEntry::InputType::Vector:
+            {
                 vector = va_arg(args, Vector *);
-                for (int h = 0; h < vector->nirrep(); ++h) {
-                    for (int row = 0; row < vector->dimpi()[h]; ++row) {
-                        *arrayPtr++ = vector->get(h, row);
-                    }
-                }
+                auto size = vector->dimpi().sum();
+                std::copy(vector->pointer(), vector->pointer() + size, arrayPtr);
```

Comment:
This copy is dependent on how the storage in Vector is done. If (for some reason) Vector's implementation changes this could fail.

Reply:
Change how? Store the data from a single irrep non-contiguously? Store the data from one irrep not contiguously to another irrep?

The way I see it, _some_ dependence on how storage is done in the `Vector` class is unavoidable, unless I swap out `arrayPtr` for an actual `Vector` object.